In [15]:
from PIL import Image
import glob
import os
import numpy as np

import tensorflow as tf
from tensorflow import keras

## 모델 변수 설정

In [77]:
# 리사이즈 변수
resize_num = 112

# 모델 설계 변수 설정
Conv2D_1 = 64
Conv2D_2 = 128
Conv2D_3 = 512
dense = 256

kernel_size = (3,3)
maxpool_size = (2,2)

# 모델 학습 변수 설정
epochs_nums = 20

### 파일 리사이즈 하는 함수

In [53]:
 # 파일 리사이즈 하는 함수
def resize_images(img_path, resize_number):
    images=glob.glob(img_path + "/*.jpg")  
    
    print(len(images), " images to be resized.")

    # 파일마다 모두 28x28 사이즈로 바꾸어 저장합니다.
    target_size=(resize_number, resize_number)
    save_path = img_path + "/resize" + str(resize_number)
    
    if not os.path.exists(save_path):
        os.mkdir(save_path)
    
    for img in images:
        old_img=Image.open(img)
        new_img=old_img.resize(target_size,Image.ANTIALIAS)
        new_img.save(save_path + "/" + img.split("/")[-1], "JPEG")
    
    print(len(images), " images resized.")
    
    return len(images)

# /aiffel/바꿀폴더/scissor or rock or paper 바꾸기
image_dir_path = os.getenv("HOME") + "/aiffel/aiffel_rsp/scissor"
image_count = 0;

image_count += resize_images(image_dir_path, resize_num)
print("가위 이미지 resize 완료!")

image_dir_path = os.getenv("HOME") + "/aiffel/aiffel_rsp/rock"
image_count += resize_images(image_dir_path, resize_num)
print("바위 이미지 resize 완료!")

image_dir_path = os.getenv("HOME") + "/aiffel/aiffel_rsp/paper"
image_count += resize_images(image_dir_path, resize_num)
print("보 이미지 resize 완료!")

print(image_count)

# test set
image_dir_path = os.getenv("HOME") + "/aiffel/aiffel_rsp/test/scissor"
image_count = 0;

image_count += resize_images(image_dir_path, resize_num)
print("가위 이미지 resize 완료!")

image_dir_path = os.getenv("HOME") + "/aiffel/aiffel_rsp/test/rock"
image_count += resize_images(image_dir_path, resize_num)
print("바위 이미지 resize 완료!")

image_dir_path = os.getenv("HOME") + "/aiffel/aiffel_rsp/test/paper"
image_count += resize_images(image_dir_path, resize_num)
print("보 이미지 resize 완료!")

print(image_count)

100  images to be resized.
100  images resized.
가위 이미지 resize 완료!
100  images to be resized.
100  images resized.
바위 이미지 resize 완료!
100  images to be resized.
100  images resized.
보 이미지 resize 완료!
300
100  images to be resized.
100  images resized.
가위 이미지 resize 완료!
100  images to be resized.
100  images resized.
바위 이미지 resize 완료!
100  images to be resized.
100  images resized.
보 이미지 resize 완료!
300


### 이미지 부풀리기

In [54]:
def image_generator (img_path):
    
    save_path = img_path
    images=glob.glob(img_path + "/*.jpg")
        
    count = 0
    
    for img in images:
        old_img = Image.open(img)
        new_img = old_img.transpose(Image.FLIP_LEFT_RIGHT)
        new_img.save(save_path + "/" + str(len(images)) + str(count) + ".jpg", "JPEG")
        count += 1
        
    for img in images:
        old_img = Image.open(img)
        new_img = old_img.transpose(Image.FLIP_TOP_BOTTOM)
        new_img.save(save_path + "/" + str(len(images)) + str(count) + ".jpg", "JPEG")
        count += 1


# 이미지 경로
image_dir_path = os.getenv("HOME") + "/aiffel/aiffel_rsp/scissor/resize112"
image_generator(image_dir_path)
print("가위 부풀리기 완료!")

image_dir_path = os.getenv("HOME") + "/aiffel/aiffel_rsp/rock/resize112"
image_generator(image_dir_path)
print("바위 부풀리기 완료!")

image_dir_path = os.getenv("HOME") + "/aiffel/aiffel_rsp/paper/resize112"
image_generator(image_dir_path)
print("보 부풀리기 완료!")

가위 부풀리기 완료!
바위 부풀리기 완료!
보 부풀리기 완료!
300


In [60]:
image_count = 0;

image_dir_path = os.getenv("HOME") + "/aiffel/aiffel_rsp/scissor/resize112"
images=glob.glob(image_dir_path + "/*.jpg")
image_count += len(images)

image_dir_path = os.getenv("HOME") + "/aiffel/aiffel_rsp/rock/resize112"
images=glob.glob(image_dir_path + "/*.jpg")
image_count += len(images)

image_dir_path = os.getenv("HOME") + "/aiffel/aiffel_rsp/paper/resize112"
images=glob.glob(image_dir_path + "/*.jpg")
image_count += len(images)

print(image_count)

900


### 이미지 라벨링 하는 함수
### 가위 0, 바위 1, 보 2

In [61]:
# 이미지 라벨링 하는 함수
def load_data(img_path, number_of_data, param_img_size):  # 가위바위보 이미지 개수 총합에 주의하세요.
    # 가위 : 0, 바위 : 1, 보 : 2
    img_size=param_img_size
    color=3
    #이미지 데이터와 라벨(가위 : 0, 바위 : 1, 보 : 2) 데이터를 담을 행렬(matrix) 영역을 생성합니다.
    imgs=np.zeros(number_of_data*img_size*img_size*color,dtype=np.int32).reshape(number_of_data,img_size,img_size,color)
    labels=np.zeros(number_of_data,dtype=np.int32)

    print(img_path + '/scissor/resize' + str(img_size) + '/*.jpg')
    
    idx=0
    for file in glob.iglob(img_path + '/scissor/resize' + str(img_size) + '/*.jpg'):
        img = np.array(Image.open(file),dtype=np.int32)
        imgs[idx,:,:,:]=img    # 데이터 영역에 이미지 행렬을 복사
        labels[idx]=0   # 가위 : 0
        idx=idx+1

    for file in glob.iglob(img_path+'/rock/resize' + str(img_size) + '/*.jpg'):
        img = np.array(Image.open(file),dtype=np.int32)
        imgs[idx,:,:,:]=img    # 데이터 영역에 이미지 행렬을 복사
        labels[idx]=1   # 바위 : 1
        idx=idx+1  
    
    for file in glob.iglob(img_path+'/paper/resize' + str(img_size) + '/*.jpg'):
        img = np.array(Image.open(file),dtype=np.int32)
        imgs[idx,:,:,:]=img    # 데이터 영역에 이미지 행렬을 복사
        labels[idx]=2   # 보 : 2
        idx=idx+1
        
    print("학습데이터(x_train)의 이미지 개수는", idx,"입니다.")
    return imgs, labels

# /aiffel/바꿔야하는 폴더
image_dir_path = os.getenv("HOME") + "/aiffel/aiffel_rsp"
(x_train, y_train)=load_data(image_dir_path,image_count, resize_num)
x_train_norm = x_train/255.0   # 입력은 0~1 사이의 값으로 정규화

print("x_train shape: {}".format(x_train.shape))
print("y_train shape: {}".format(y_train.shape))

/aiffel/aiffel/aiffel_rsp/scissor/resize112/*.jpg
학습데이터(x_train)의 이미지 개수는 900 입니다.
x_train shape: (900, 112, 112, 3)
y_train shape: (900,)


### 모델 설계

In [79]:
# model을 직접 만들어 보세요.
# Hint! model의 입력/출력부에 특히 유의해 주세요. 가위바위보 데이터셋은 MNIST 데이터셋과 어떤 점이 달라졌나요?

model=keras.models.Sequential()
model.add(keras.layers.Conv2D(Conv2D_1, kernel_size, activation='relu', input_shape=(resize_num,resize_num,3)))
model.add(keras.layers.MaxPool2D(maxpool_size))
model.add(keras.layers.Conv2D(Conv2D_2, kernel_size, activation='relu'))
model.add(keras.layers.MaxPooling2D(maxpool_size))
model.add(keras.layers.Conv2D(Conv2D_3, kernel_size, activation='relu'))
model.add(keras.layers.MaxPooling2D(maxpool_size))
model.add(keras.layers.Flatten())
model.add(keras.layers.Dense(dense, activation='relu'))
model.add(keras.layers.Dense(3, activation='softmax'))

print('Model에 추가된 Layer 개수: ', len(model.layers))

# 모델 요약
model.summary()

Model에 추가된 Layer 개수:  9
Model: "sequential_9"
_________________________________________________________________
Layer (type)                 Output Shape              Param #   
conv2d_24 (Conv2D)           (None, 110, 110, 64)      1792      
_________________________________________________________________
max_pooling2d_24 (MaxPooling (None, 55, 55, 64)        0         
_________________________________________________________________
conv2d_25 (Conv2D)           (None, 53, 53, 128)       73856     
_________________________________________________________________
max_pooling2d_25 (MaxPooling (None, 26, 26, 128)       0         
_________________________________________________________________
conv2d_26 (Conv2D)           (None, 24, 24, 512)       590336    
_________________________________________________________________
max_pooling2d_26 (MaxPooling (None, 12, 12, 512)       0         
_________________________________________________________________
flatten_9 (Flatten)          (

### 모델 학습

In [80]:
# model을 학습시키는 코드를 직접 작성해 보세요.
# Hint! model.compile()과 model.fit()을 사용해 봅시다.
model.compile(optimizer='adam',
             loss='sparse_categorical_crossentropy',
             metrics=['accuracy'])

model.fit(x_train, y_train, epochs=epochs_nums)

Epoch 1/20
29/29 [==============================] - 2s 35ms/step - loss: 62.7510 - accuracy: 0.3233
Epoch 2/20
29/29 [==============================] - 1s 28ms/step - loss: 1.0939 - accuracy: 0.3500
Epoch 3/20
29/29 [==============================] - 1s 28ms/step - loss: 1.0695 - accuracy: 0.4667
Epoch 4/20
29/29 [==============================] - 1s 28ms/step - loss: 1.0207 - accuracy: 0.5467
Epoch 5/20
29/29 [==============================] - 1s 28ms/step - loss: 0.9812 - accuracy: 0.5100
Epoch 6/20
29/29 [==============================] - 1s 28ms/step - loss: 0.8568 - accuracy: 0.5889
Epoch 7/20
29/29 [==============================] - 1s 28ms/step - loss: 0.7215 - accuracy: 0.6700
Epoch 8/20
29/29 [==============================] - 1s 28ms/step - loss: 0.6175 - accuracy: 0.7467
Epoch 9/20
29/29 [==============================] - 1s 28ms/step - loss: 0.6618 - accuracy: 0.7300
Epoch 10/20
29/29 [==============================] - 1s 28ms/step - loss: 0.5294 - accuracy: 0.7900
Epoch 11

### 트레이닝 셋 만들기

In [68]:
# x_test, y_test를 만드는 방법은 x_train, y_train을 만드는 방법과 아주 유사합니다.

image_dir_path = os.getenv("HOME") + "/aiffel/aiffel_rsp/test"
print(image_dir_path)
(x_test, y_test)=load_data(image_dir_path, 300, resize_num)
x_test_norm = x_train/255.0   # 입력은 0~1 사이의 값으로 정규화

print("x_test shape: {}".format(x_test.shape))
print("y_test shape: {}".format(y_test.shape))

/aiffel/aiffel/aiffel_rsp/test
/aiffel/aiffel/aiffel_rsp/test/scissor/resize112/*.jpg
학습데이터(x_train)의 이미지 개수는 300 입니다.
x_test shape: (300, 112, 112, 3)
y_test shape: (300,)


### 테스트 결과

In [81]:
# model을 학습시키는 코드를 직접 작성해 보세요.
# Hint! model.evaluate()을 사용해 봅시다.
test_loss, test_accuracy = model.evaluate(x_test,y_test, verbose=2)
print("test_loss: {} ".format(test_loss))
print("test_accuracy: {}".format(test_accuracy))

10/10 - 0s - loss: 3.0801 - accuracy: 0.7600
test_loss: 3.0801265239715576 
test_accuracy: 0.7599999904632568


In [93]:
predicted_result = model.predict(x_test)  # model이 추론한 확률값. 
predicted_labels = np.argmax(predicted_result, axis=1)

wrong_predict_list=[]
for i, _ in enumerate(predicted_labels):
    if predicted_labels[i] != y_test[i]:
        wrong_predict_list.append(i)

zero_count = 0
one_count = 0
two_count = 0

for n in wrong_predict_list:
    print("예측확률분포: " + str(predicted_result[n]))
    print("라벨: " + str(y_test[n]) + ", 예측결과: " + str(predicted_labels[n]))
    if str(y_test[n]) == '0' :
        zero_count += 1
    elif str(y_test[n]) == '1':
        one_count += 1
    elif str(y_test[n]) == '2':
        two_count += 1
        
print("가위 : " + str(zero_count) + ", 바위 : " + str(one_count) + ", 보 : " + str(two_count))

예측확률분포: [4.6363512e-01 1.8406344e-12 5.3636491e-01]
라벨: 0, 예측결과: 2
예측확률분포: [1.8605009e-02 6.1053827e-12 9.8139495e-01]
라벨: 0, 예측결과: 2
예측확률분포: [4.2429030e-01 2.9905259e-17 5.7570964e-01]
라벨: 0, 예측결과: 2
예측확률분포: [2.3730168e-02 1.4963361e-16 9.7626978e-01]
라벨: 0, 예측결과: 2
예측확률분포: [1.0687573e-06 3.2899392e-34 9.9999893e-01]
라벨: 0, 예측결과: 2
예측확률분포: [2.0445584e-01 3.4525011e-13 7.9554415e-01]
라벨: 0, 예측결과: 2
예측확률분포: [2.2061105e-01 5.3844554e-13 7.7938896e-01]
라벨: 0, 예측결과: 2
예측확률분포: [5.6490112e-06 1.5277713e-17 9.9999440e-01]
라벨: 0, 예측결과: 2
예측확률분포: [7.102432e-02 4.650432e-17 9.289757e-01]
라벨: 0, 예측결과: 2
예측확률분포: [1.0418311e-05 0.0000000e+00 9.9998963e-01]
라벨: 0, 예측결과: 2
예측확률분포: [0.25290155 0.         0.74709845]
라벨: 0, 예측결과: 2
예측확률분포: [1.0507248e-01 6.2589997e-17 8.9492744e-01]
라벨: 0, 예측결과: 2
예측확률분포: [4.0247562e-04 2.3236551e-15 9.9959749e-01]
라벨: 0, 예측결과: 2
예측확률분포: [2.8545477e-05 2.9913863e-35 9.9997151e-01]
라벨: 0, 예측결과: 2
예측확률분포: [9.5107220e-07 8.9999513e-37 9.9999905e-01]
라벨: 0, 예측결과: 2
예측확률분포:

### 테스트 기록

### 1. 이미지 개수 300, 이미지 사이즈 28, Conv2D(64,128,256), Dense(128), epoch=10
train accuracy = 0.84
test accuracy : 0.54

### 2. 이미지 개수 300, 이미지 사이즈 56, Conv2D(64,128,256), Dense(128), epoch=10
train accuracy = 0.99
test accuracy : 0.43

### 3. 이미지 개수 300, 이미지 사이즈 112, Conv2D(64,128,256), Dense(128), epoch=10
train accuracy = 0.72
test accuracy : 0.31

이미지가 커지면서 정확도가 증가할줄 알았는데, 사이즈를 늘릴때 줄어든거보니 epoch이 부족했던건 아닐까 생각이 들어 늘려보았다

### 4. 이미지 개수 300, 이미지 사이즈 112, Conv2D(64,128,256), Dense(128), epoch=15
train accuracy = 0.98
test accuracy : 0.36

### 5. 이미지 개수 300, 이미지 사이즈 112, Conv2D(64,128,256), Dense(128), epoch=20
train accuracy = 1
test accuracy : 0.37

학습을 하는 중간에 1을 달성하는것을 보니 무작정 epoch를 늘리는것은 또 좋지 않은듯하다.
이미지 사이즈 112, epoch는 20으로 고정하고 하이퍼 파라미터를 늘려보았다.

### 6. 이미지 개수 300, 이미지 사이즈 112, Conv2D(128,256,512), Dense(128), epoch=20
train accuracy = 0.98
test accuracy : 0.45

### 7. 이미지 개수 300, 이미지 사이즈 112, Conv2D(128,256,512), Dense(256), epoch=20
train accuracy = 0.97
test accuracy : 0.51

커널 사이즈를 변경해볼까?
kernel_size = (3,3) -> (4,4)
maxpool_size = (2,2) -> (3,3)

### 8. 이미지 개수 300, 이미지 사이즈 112, Conv2D(128,256,512), Dense(256), epoch=20
train accuracy = 0.97
test accuracy : 0.37

커널 사이즈 변경은 아직인듯 하다. 다시 kernel_size = (3,3) maxpool_size = (2,2)으로 변경
이미지의 갯수가 좀 부족한가?
Image의 내장함수인 transpose로 좌우, 상하 반전을 시켜서 이미지 증가

### 9. 이미지 개수 900, 이미지 사이즈 112, Conv2D(128,256,512), Dense(256), epoch=20
train accuracy = 0.97
test accuracy : 0.75

### 회고록

처음 시작할때는 팀원들의 데이터를 받아서 진행을 했었다.   
내 이미지로 학습을 한다음에 팀원들의 이미지로 테스트를 해보니 학습 정확도는 높은데 테스트 정확도가 낮게 나왔다.  
데이터를 살펴보니 손이 정확하게 구별이 안되게 배경에 옷이나 얼굴이 포함된 데이터가 몇몇 있어서, 구별을 잘 못하나 생각을 했었다.  
그래서 웹캠이 없을때 자료를 준다는 lms 글이 생각나서 일단 그 자료를 사용해 보기로 했다.  
매번 사이즈를 바꿀때마다 원본이 덮어 씌어지는게 안좋은듯해서 원본 데이터와 resize한 사진을 분리했다.  
그런 다음 이미지의 사이즈 별로, 학습해보고, 하이퍼 파라미터를 바꾸거나, epoch의 횟수를 변경하면서 학습을 해봤다.  
이러한 변수들을 변경하면서 정확도가 늘어나긴 했지만 가끔은 떨어지기도 했다.  
그래서 데이터가 부족한것 같다는 생각이 들어 image패키지의 transpose함수를 이용해 좌우 상하 반전을 해서 데이터의 양을 늘렸다.  
학습 데이터가 300 > 900개로 증가해서 그런지 정확도가 생각보다 많이 늘어나서 목표했던 0.6을 넘길 수 있었다.  
이번 exploration을 진행하면서 학습을 하기전 데이터의 양이 중요할뿐만 아니라, 상하좌우반전이라던지 같은 사진이라도 다른 데이터가 될 수 있다는걸 깨달았다.  
모델의 하이퍼 파라미터 값이라던지, 모델이 정확하게 어떻게 동작하는지 알고 사용했더라면 좀더 이미지를 학습하는데 도움이 될것 같다는 생각이 들었다.마지막으로 데이터를 좀더 다양한 배경이나 손을 가지고 학습을 해볼것이다.  
